Make user and group optional

nickptrvc opened the pull request:
fixes #225

- Refactor resources/providers
- Add service startup check
- Add additional startup action
- Add additional properties to install resource

iennae commented:
@nickptrvc 

Can you update your resource usage to use the new 12.5 model as shown in https://github.com/chef-cookbooks/tomcat/blob/master/resources/install.rb? Your refactor is using the older separate resource and provider files which we are moving away from. We can't use this PR as is. Thanks for understanding.

nickptrvc commented:
@iennae 
I fairly new to Chef. Can you show me any documentation on declaring resources the way you're asking? I didn't know it had changed and wasn't able to find examples of it being declared that way. 

And just to be clear, do you mean this part...
```ruby
action_class do
# ...
end
```

iennae commented:
@nickptrvc Sure, totally understand being new to it. Check out the docs site here: https://docs.chef.io/custom_resources.html. If you look at the current master branch for the file https://github.com/chef-cookbooks/tomcat/blob/master/resources/install.rb this is a good example in this cookbook. You'll see that only one file needs to be created now, not separate resource/provider files. Within the resource file, you'll find both the attributes for the resource defined with the keyword **property** and actions defined with **action**.

nickptrvc commented:
@iennae 
I reimplemented the feature by using your suggestions with defining the properties. I also simplified the PR by not including the option to create a service when using the tomcat_install resource. Its probably best to keep the PR simple for now.

In regards to the documentation, the only thing I could find on `action_class` was that it should not be used because it is considered part of the private api. If you have something different, could you please send me the link? I'd like to have a better understanding of what I should and should not use. 

> This method is part of a private API. You should avoid using this method if possible, as it may be removed or be changed in the future.
http://www.rubydoc.info/gems/chef/Chef%2FResource.action_class

nickptrvc commented:
fixes #225

- added tomcat_user and tomcat_group properties on tomcat_install and tomcat_service resources; properties default to tomcat_#{instance_name}
- added user and group properties to service_*.rb resources
- added user and group to service_*.erb templates
- added an only if statement to start systemd service if java was available
- fixed a typo in service_sysv_init.rb where the service name was using `instance_name` instead of `new_resource.instance_name`

nickptrvc commented:
@gbarnes619 
I made a few changes. I took out the part where you can install a service through the tomcat_install resource. I think thats something that can be revisited at another time. 

The `tomcat_install` and `tomcat_service` resources can now accept tomcat_user and tomcat_group properties. They default to the original implementation for backwards compatibility.

DigitalSmile commented:
Any target dates, when It could be merged? We need this feature so badly...

mrmarbury commented:
+1 on the "this feature is needed" part

maassjo commented:
+1 on the "this Feature is needed badly" part. We are currently in the process of automating a whole IT department, and we desperately need this feature. Please be so kind as to merge it asap.

Thanks a lot in advance!

tas50 commented:
Sorry for taking so long to get this one merged in.  I'll get it released today.
